File: frontend/src/lib/api/areas.ts
/**
 * Áreas responsáveis de etapa (mudança #3) — endpoints exclusivos do
 * AreaResponsavelPicker. Envelope `{ok, data}` desempacotado por `client.ts`.
 */

import { get, post } from './client';
import type { EtapaDetail, EtapaResponsavelArea } from '$lib/types/projectDetail';

export interface AreaOption {
	id: number;
	sigla: string;
	nome: string;
}

/** GET /api/areas — TODAS as áreas ativas (busca é client-side). */
export function fetchAreas(signal?: AbortSignal): Promise<{ areas: AreaOption[] }> {
	return get<{ areas: AreaOption[] }>('/api/areas', signal);
}

/** POST /api/etapas/<id>/responsaveis — substitui a lista completa. */
export function saveEtapaResponsaveis(
	etapaId: number,
	areas: EtapaResponsavelArea[],
	signal?: AbortSignal
): Promise<{ etapa: EtapaDetail }> {
	return post<{ etapa: EtapaDetail }>(`/api/etapas/${etapaId}/responsaveis`, { areas }, signal);
}

File: routes/api/areas.py
"""Endpoint JSON com TODAS as áreas ativas (picker de responsáveis de etapa).

Diferente de ``GET /api/orgaos/escopo`` (escopado por subtree) e de
``GET /api/admin/orgaos/opcoes`` (admin-only), este endpoint lista o catálogo
COMPLETO de ``OrgaoUnidade`` ativas para qualquer usuário logado — a busca é
100% client-side no ``AreaResponsavelPicker`` (mesmo padrão do AssigneePicker).

Anexa ao ``main_bp`` ÚNICO; NÃO cria blueprint novo.

Payload inclui ``pai_id`` para permitir montagem de árvore client-side
(mesmo padrão de ``serialize_orgao_option``, usado pelo picker de Responsável
da etapa).
"""

from __future__ import annotations

from flask import Response

from models import OrgaoUnidade

from ..blueprint import main_bp
from .envelope import ok
from .negotiation import api_login_required


@main_bp.route("/api/areas", methods=["GET"])
@api_login_required
def api_areas_listar() -> Response | tuple[Response, int]:
    """Lista TODAS as áreas (OrgaoUnidade) ativas, ordenadas por sigla.

    Returns:
        ``ok({"areas": [{"id", "sigla", "nome", "pai_id"}, ...]})`` (200); 401 sem sessão.
    """
    areas = (
        OrgaoUnidade.query.filter(OrgaoUnidade.ativo.is_(True))
        .order_by(OrgaoUnidade.sigla)
        .all()
    )
    return ok(
        {
            "areas": [
                {"id": o.id, "sigla": o.sigla, "nome": o.nome, "pai_id": o.pai_id}
                for o in areas
            ]
        }
    )
